### PR TITLE
[Auth] - Generic jwksEndpoint on Neo4jGraphQLAuthJWKSPlugin and Generic Secret on Neo4jGraphQLAuthJWTPlugin

### PR DESCRIPTION
**Pull request by farhadnowzari**
# Description
To solve the issue, both `Neo4jGraphQLAuthJWKSPlugin` and `Neo4jGraphQLAuthJWTPlugin` are updated.

we can now pass a function to `jwksEndpoint` and this function will always run by a method called `tryToResolveKeys`. When the `jwksEndpoint` is a function, the `JwksClient` will be null and gets initiated by running `jwksEndpoint` in `wrapper.ts`.

This makes the endpoint to be flexible and the host service can change it in runtime based on the incoming request. 

**Note:** Passing only a string is still supported.

The above is also applied when we pass the secret. We can pass a function to the `Secret` in `Neo4jGraphQLAuthJWTPlugin` and have it computed in runtime.

# Issue

Closes https://github.com/neo4j/graphql/issues/2333

# Checklist

The following requirements should have been met (depending on the changes in the branch):

- [x] Documentation has been updated
- [ ] TCK tests have been updated
- [ ] Integration tests have been updated
- [ ] Example applications have been updated
- [ ] New files have copyright header
- [x] CLA (https://neo4j.com/developer/cla/) has been signed

### CODE REVIEW

**Comment by darrellwarde** [comment]
Hey @farhadnowzari, sorry we missed this being opened! If you haven't, we need you to sign our CLA before we can consider merging this: https://neo4j.com/developer/cla/#sign-cla

Thanks!

**Comment by farhadnowzari** [comment]
Hi @darrellwarde 
is it possible I change the account who should sign the CLA to this one I'm commenting with?
I have opened the issue with this account and pushed with the company's account ^^

**Comment by tbwiss** [comment]
Hi @farhadnowzari, thanks for signing the CLA!
Unfortunately, you then have to sign the CLA with both your private and company account. I'm very sorry for the inconvenience! It's what our terms and legal group require us and our contributors to do.
Thank you!

**Comment by farhadnowzari** [comment]
> Hi @farhadnowzari, thanks for signing the CLA! Unfortunately, you then have to sign the CLA with both your private and company account. I'm very sorry for the inconvenience! It's what our terms and legal group require us and our contributors to do. Thank you!

Hi @tbwiss ,

all good, I just wanted to wait until today to send the company's CLA because I wanted to do it with the company's notebook :)

but now both CLAs should be there.

Cheers

**Comment by tbwiss** [comment]
Hi @farhadnowzari,
your CLA check is now passing which is great!

You do however have some conflicts now which need to be resolved before the test suite is executed.

On the same note, it seems like you have different formatting settings for your code. See for instance `packages/graphql/src/types.ts`. I assume it's prettier causing this.

**Comment by farhadnowzari** [comment]
> Hi @farhadnowzari, your CLA check is now passing which is great!
> 
> You do however have some conflicts now which need to be resolved before the test suite is executed.
> 
> On the same note, it seems like you have different formatting settings for your code. See for instance `packages/graphql/src/types.ts`. I assume it's prettier causing this.

Hi @tbwiss 

I see, I will resolve the conflicts.

**Comment by farhadnowzari** [comment]
> > Hi @farhadnowzari, your CLA check is now passing which is great!
> > You do however have some conflicts now which need to be resolved before the test suite is executed.
> > On the same note, it seems like you have different formatting settings for your code. See for instance `packages/graphql/src/types.ts`. I assume it's prettier causing this.
> 
> Hi @tbwiss
> 
> Yes that's by prettier.
> 
> I see, I will resolve the conflicts.

Hi @tbwiss 

I resolved the conflicts yesterday, I think this should be good to go.
Let me know if there is anything else I can do.
Thanks :)

**Comment by neo4j-team-graphql** [comment]
Made some changes to the GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-stage-2359.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

**Comment by farhadnowzari** [comment]
Not really, my changes are only on the auth plugin.I just merged dev into mineSent from my iPhoneOn 21. Dec 2022, at 10:52, Neo4j Team GraphQL ***@***.***> wrote:﻿
Made some changes to the GraphQL Toolbox? Or made some changes to the core library perhaps?
Check out your changes at https://neo4j-graphql-toolbox-stage-2359.surge.sh
Use for instance a local dbms hosted on Neo4j Desktop or a cloud-hosted Aura Free dbms.

**Comment by farhadnowzari** [comment]
Hi @tbwiss 

Merry Christmas.

I have a question regarding the changes under this PR.
I got a mail that said if I have done any changes to the core library. I didn't, I just merged the dev to my branch to resolve the conflicts.

is there anything I can help with to make this PR released?
At the moment I'm using the changes locally but that can get tricky to maintain due to future changes on the library.

Thanks in advance.

Happy Holidays

Kind Regards
Farhad

File: packages/plugins/graphql-plugin-auth/README.md [hunk 1]
```diff
@@ -26,6 +26,19 @@ const neoSchema = new Neo4jGraphQL({
         }),
     },
 });
+
```
Comment (tbwiss) [inline packages/plugins/graphql-plugin-auth/README.md:29]: Thanks for adding this info to the README.md.
Let's also add this to the Documentation in the setup page for Auth: https://github.com/neo4j/graphql/blob/dev/docs/modules/ROOT/pages/auth/setup.adoc

File: packages/plugins/graphql-plugin-auth/src/Neo4jGraphQLAuthJWTPlugin.ts [hunk 2]
```diff
@@ -62,7 +75,7 @@ class Neo4jGraphQLAuthJWTPlugin {
 
                 result = jsonwebtoken.decode(token, { json: true }) as unknown as T;
             }
-
+            //In case that the secret is null, make sure the 'tryToResolveKeys' is ran before the decode method.
```
Comment (tbwiss) [inline packages/plugins/graphql-plugin-auth/src/Neo4jGraphQLAuthJWTPlugin.ts:78]: A similar text could be used as a `debug()` output if `this.secret === null` I suppose? Along the line of what you did in the packages/plugins/graphql-plugin-auth/src/Neo4jGraphQLAuthJWKSPlugin.ts file.

File: packages/plugins/graphql-plugin-auth/src/Neo4jGraphQLAuthJWTPlugin.test.ts [hunk 3]
```diff
@@ -39,6 +39,30 @@ describe("Neo4jGraphQLAuthJWTPlugin", () => {
         expect(decoded).toMatchObject(payload);
     });
 
+    test("should decode token when secret is a generic", async () => {
```
Comment (tbwiss) [inline packages/plugins/graphql-plugin-auth/src/Neo4jGraphQLAuthJWTPlugin.test.ts:42]: Thanks for the test case!
Would you be able to add another test case if the `secret` function returns `null` or `undefined`?
Similar to:
```javascript
secret: () => {
    return null;
},
```

File: packages/plugins/graphql-plugin-auth/src/Neo4jGraphQLAuthJWKSPlugin.ts [hunk 4]
```diff
@@ -82,6 +109,7 @@ class Neo4jGraphQLAuthJWKSPlugin {
         };
 
         return new Promise((resolve, reject) => {
+            if (!this.client) reject("JwksClient should NOT be empty! Make sure the 'tryToResolveKeys' method is called before decoding");
```
Comment (tbwiss) [inline packages/plugins/graphql-plugin-auth/src/Neo4jGraphQLAuthJWKSPlugin.ts:112]:
```suggestion
            if (!this.client) reject("JwksClient should not be empty! Make sure the 'tryToResolveKeys' method is called before decoding");
```

File: packages/graphql/src/schema/resolvers/wrapper.ts [hunk 5]
```diff
@@ -32,6 +32,7 @@ import { getToken, parseBearerToken } from "../../utils/get-token";
 import type { SubscriptionConnectionContext, SubscriptionContext } from "./subscriptions/types";
 import { decodeToken, verifyGlobalAuthentication } from "./wrapper-utils";
 import type { Neo4jGraphQLSchemaModel } from "../../schema-model/Neo4jGraphQLSchemaModel";
+import { IncomingMessage } from "http";
```
Comment (tbwiss) [inline packages/graphql/src/schema/resolvers/wrapper.ts:35]: Would you be able to format this file with `prettier` and the prettier config provided in this repo? thanks!

File: packages/graphql/src/schema/resolvers/wrapper.ts [hunk 6]
```diff
@@ -86,69 +87,72 @@ export const wrapResolver =
...
         context.callbacks = config.callbacks;
 
-        if (!context.jwt) {
-            const token = getToken(context);
-            context.jwt = await decodeToken(token, context.plugins.auth);
-        }
+                if (!context.jwt) {
+                    if (context.plugins.auth)
+                        //Here we will try to compute the generic Secret or the generic jwksEndpoint
+                        context.plugins.auth.tryToResolveKeys(context instanceof IncomingMessage ? context : context.req || context.request);
```
Comment (tbwiss) [inline packages/graphql/src/schema/resolvers/wrapper.ts:93]:
```suggestion
                    if (context.plugins.auth) {
                         // Here we will try to compute the generic Secret or the generic jwksEndpoint
                        const contextRequest = context.req || context.request
                        context.plugins.auth.tryToResolveKeys(context instanceof IncomingMessage ? context : contextRequest);
                    }
```

**Review by tbwiss — COMMENTED** [review]
Hi @farhadnowzari!
Sorry for the delay on this. I left some comments. Some are a little bit nitpicky, I hope you don't mind.
Please also make sure to run `prettier` with the config of this repo on all the files that you made changes. For instance the packages/graphql/src/schema/resolvers/wrapper.ts has a lot of indentation changes which are not required.

File: packages/plugins/graphql-plugin-auth/README.md (hunk 1 above)
Comment (farhadnowzari) [reply packages/plugins/graphql-plugin-auth/README.md:29]: resolved by 7ccc4ae3555a648b572aa4db8f5e04d58f5ec51b

File: packages/plugins/graphql-plugin-auth/src/Neo4jGraphQLAuthJWTPlugin.ts (hunk 2 above)
Comment (farhadnowzari) [reply packages/plugins/graphql-plugin-auth/src/Neo4jGraphQLAuthJWTPlugin.ts]: resolved by 90fc0a0c7fa7c4fdeeacfb1ed13b4eaef98853fd

Comment (farhadnowzari) [reply packages/plugins/graphql-plugin-auth/src/Neo4jGraphQLAuthJWTPlugin.ts]: btw, the `debug.enabled` is always false so on the plugins the `debug()` would not do anything, would it? I found it out because I wanted to create a test case when `secret is null` to check if debug is called or `console.log` but it never has an output unless I set the `debug.enabled = true`

File: packages/graphql/src/schema/resolvers/wrapper.ts (hunk 5 above)
Comment (farhadnowzari) [reply packages/graphql/src/schema/resolvers/wrapper.ts:35]: The changed files are now formatted with prettier in this PR, 864d3a94be987800114d78ac156a0bd09bd12f1d

File: packages/plugins/graphql-plugin-auth/src/Neo4jGraphQLAuthJWTPlugin.test.ts (hunk 3 above)
Comment (farhadnowzari) [reply packages/plugins/graphql-plugin-auth/src/Neo4jGraphQLAuthJWTPlugin.test.ts:42]: Hi @tbwiss ,

about the `null` test case I'm not sure. 
In two cases the secret can be null.
1. The `tryToResolveKeys` is not called properly in `wrapper.ts`
2. The getter for secret is returning null.

in both cases the `decode` will return `undefined`. We can create a test for the first case to check if `debug()` is called in case the `tryToResolveKeys` is not called but the second case I think would be test case for the host application.

for example if I assign a function to `secret` from my service, I have to unit-test that function inside that service right? so it doesn't return null.

but maybe in both cases we can add a test which says if `secret` was null then check if debug is called but as I mentioned earlier the `debug()` will never be called because apparently the `debug.enabled` is always false and we are not setting it to true anywhere. 

let me know if we need to change this, then I will try to add a `Debug.Debugger` to the inputs so then we can maybe mock it

File: packages/graphql/src/schema/resolvers/wrapper.ts (hunk 5 above)
Comment (tbwiss) [reply packages/graphql/src/schema/resolvers/wrapper.ts:35]: Perfect! Thanks!

File: packages/plugins/graphql-plugin-auth/src/Neo4jGraphQLAuthJWTPlugin.test.ts (hunk 3 above)
Comment (tbwiss) [reply packages/plugins/graphql-plugin-auth/src/Neo4jGraphQLAuthJWTPlugin.test.ts:42]: Yes, I understand your point!
We can't however rely on the user/customer/service code to be unit tested. Therefore we have to expect `null` as a potential value of `secret()`.
If it's possible for you to add a test for this scenario that would be really good. Once that is in place I would say we can merge this PR.

Comment (farhadnowzari) [reply packages/plugins/graphql-plugin-auth/src/Neo4jGraphQLAuthJWTPlugin.test.ts:42]: on it 👍

Comment (farhadnowzari) [reply packages/plugins/graphql-plugin-auth/src/Neo4jGraphQLAuthJWTPlugin.test.ts:41]: Hi @tbwiss 

I added a test but it is not really the case where the computed value is null, that value can't be null since the function must return only type `Secret`.

image

but I now check if the `secret` is passed as a function and if the result value is still null, it throws an exception.

Comment (tbwiss) [reply packages/plugins/graphql-plugin-auth/src/Neo4jGraphQLAuthJWTPlugin.test.ts:41]: @farhadnowzari, thank you!

**Review by tbwiss — APPROVED** [review]
LGTM now!
Thank you very much @farhadnowzari for your contribution!

**Comment by farhadnowzari** [comment]
> LGTM now!
> Thank you very much @farhadnowzari for your contribution!

🙏🏼😊

**Comment by tbwiss** [comment]
@farhadnowzari were you meant to close the PR without it being merged?

**Comment by farhadnowzari** [comment]
> @farhadnowzari were you meant to close the PR without it being merged?

Oh, no 😅 that was by mistake.
Let me know if there is anything else I can help with
